cpp/src/parquet/chunker_internal.h

```diff
@@ -0,0 +1,166 @@
+/// chunks.
+///
+/// Implementation details:
+///
+/// Only the parquet writer must be aware of the content defined chunking, the reader
+/// doesn't need to know about it. Each parquet column writer holds a
+/// ContentDefinedChunker instance depending on the writer's properties. The chunker's
+/// state is maintained across the entire column without being reset between pages and row
+/// groups.
+///
+/// The chunker receives the record shredded column data (def_levels, rep_levels, values)
+/// and goes over the (def_level, rep_level, value) triplets one by one while adjusting
+/// the column-global rolling hash based on the triplet. Whenever the rolling hash matches
+/// a predefined mask, the chunker creates a new chunk. The chunker returns a vector of
+/// Chunk objects that represent the boundaries of the chunks.
+/// Note that the boundaries are deterministically calculated exclusively based on the
+/// data itself, so the same data will always produce the same chunks - given the same
+/// chunker configuration.
+///
+/// References:
+/// - FastCDC: a Fast and Efficient Content-Defined Chunking Approach for Data
+///   Deduplication
+///   https://www.usenix.org/system/files/conference/atc16/atc16-paper-xia.pdf
+/// - Git is for Data (chunk size normalization used here is described in section 6.2.1):
+///   https://www.cidrdb.org/cidr2023/papers/p43-low.pdf
+class ContentDefinedChunker {
+ public:
+  /// Create a new ContentDefinedChunker instance
+  ///
+  /// @param level_info Information about definition and repetition levels
+  /// @param size_range Min/max chunk size as pair<min_size, max_size>, the chunker will
+  ///                   attempt to uniformly distribute the chunks between these extremes.
+  /// @param norm_factor Normalization factor to center the chunk size around the average
+  ///                    size more aggressively. By increasing the normalization factor,
+  ///                    probability of finding a chunk boundary increases.
+  ContentDefinedChunker(const LevelInfo& level_info,
+                        std::pair<uint64_t, uint64_t> size_range,
+                        uint8_t norm_factor = 0);
+
+  /// Get the chunk boundaries for the given column data
+  ///
+  /// @param def_levels Definition levels
+  /// @param rep_levels Repetition levels
+  /// @param num_levels Number of levels
+  /// @param values Column values as an Arrow array
+  /// @return Vector of Chunk objects representing the chunk boundaries
+  const std::vector<Chunk> GetBoundaries(const int16_t* def_levels,
+                                         const int16_t* rep_levels, int64_t num_levels,
+                                         const ::arrow::Array& values);
+
+ private:
+  // Update the rolling hash with a compile-time known sized value, set has_matched_ to
+  // true if the hash matches the mask.
+  template <typename T>
+  void Roll(const T value);
+
```

Comment:
Ok, instead of declaring private methods in the `.h`, we should use the pimpl idiom and hide the implementation entirely, e.g.:
```suggestion
 private:
  class Impl;
  std::unique_ptr<Impl> impl_;
};
```

Reply:
Updating.

Reply:
Well, everything is `internal` at the moment (both the header and the namespace), do we need to further hide the implementation?

Reply:
Updated, note that I'm planning a follow-up to unify the fixed-sized-chunking logic used at `DoInBatches` and the CDC chunker implemented here behind a common chunker API. Data page adding happens on different conditions, so it is not entirely trivial to refactor but I believe it is doable.